explorer/service/src/main/java/org/radixdlt/explorer/helper/DumpHelper.java

```diff
@@ -0,0 +1,119 @@
+package org.radixdlt.explorer.helper;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.RandomAccessFile;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.RejectedExecutionException;
+
+import static java.nio.file.StandardCopyOption.ATOMIC_MOVE;
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+
+public class DumpHelper {
+    private static final Logger LOGGER = LoggerFactory.getLogger("org.radixdlt.explorer");
+    private final ExecutorService dumpExecutor;
+
+
+    public DumpHelper() {
+        this.dumpExecutor = Executors.newSingleThreadExecutor();
+    }
+
+    /**
+     * Shuts down the dump executor, allowing the currently executing
+     * task to finish, but not starting any new tasks.
+     */
+    public void stopDumpExecutor() {
+        dumpExecutor.shutdownNow();
+    }
+
+    /**
+     * Makes an atomic (-ish) dump of the provided data to the provided
+     * path.
+     *
+     * @param data The data to dump.
+     * @param path The path to the file to dump it in (or null, in which
+     *             case nothing is dumped.
+     */
+    public void dumpData(byte[] data, Path path) {
+        if (path == null) {
+            return;
+        }
+
+        try {
+            dumpExecutor.submit(() -> {
+                // Ensure parent folder exists
+                path.toAbsolutePath().getParent().toFile().mkdirs();
+                Path tmpFilePath = path.getParent().resolve("tmp.txt").toAbsolutePath();
+                Path targetFilePath = path.toAbsolutePath();
+
+                try (FileOutputStream fileOutputStream = new FileOutputStream(tmpFilePath.toString())) {
+                    fileOutputStream.write(data);
+                    fileOutputStream.flush();
+                    fileOutputStream.getFD().sync();
+                    Files.move(tmpFilePath, targetFilePath, ATOMIC_MOVE, REPLACE_EXISTING);
```

Comment:
Can we move the file **after** the `fileOutputStream` has been closed?

Reply:
Very good point! I'll change immediately.